index.bs

```diff
@@ -535,167 +750,331 @@ To determine whether an |attribute| <dfn>contains a javascript: URL</dfn>:
-1. [=SanitizerConfig/Remove=] |element| from |configuration|["{{SanitizerConfig/elements}}"].
-1. [=list/Append=] |element| to |configuration|["{{SanitizerConfig/elements}}"].
-1. [=SanitizerConfig/Remove=] |element| from |configuration|["{{SanitizerConfig/removeElements}}"].
-1. [=SanitizerConfig/Remove=] |element| from |configuration|["{{SanitizerConfig/replaceWithChildrenElements}}"].
+<div class=note><span class=marker>Note:</span>
+This requires that we distinguish 4 cases:
+* Whether we have a global allow- or remove-list, and
+* whether they already contain |element| or not.
 
-NOTE: Handling of [=allowElement=] is a little more complicated than the other
-    methods, because the element allow list can have per-element allow- and
-    remove-attribute lists. We first remove the given element from the list
-    before then adding it, which has the effect of re-setting (rather than
-    merging or elsehow modifying) the per-element list to whatever is passed
-    in. In other words, the per-element allow- and remove-lists can only be
-    set as a whole.
+</div>
 
-NOTE: [=SanitizerConfig/Remove=] matches on name and namespace, so adding an
-    element with attributes would still remove the matching element from the
-    {{SanitizerConfig/removeElements}} and {{SanitizerConfig/replaceWithChildrenElements}} lists.
+1. Set |element| to the result of [=canonicalize a sanitizer element with attributes=] with
+    |element|.
+1. [=SanitizerConfig/Remove=] |element| from
+    |configuration|["{{SanitizerConfig/replaceWithChildrenElements}}"].
+1. If |configuration|["{{SanitizerConfig/elements}}"] [=map/exists=]:
+    1. [=Comment=]: We need to make sure the per-element attributes do not overlap with global
+        attributes.
+    1. Let |original element| be a copy of |element|.
+    1. If |element|["{{SanitizerElementNamespaceWithAttributes/attributes}}"] [=map/exists=]:
+        1. If |configuration|["{{SanitizerConfig/attributes}}"] [=map/exists=]:
+            1. Set |element|["{{SanitizerElementNamespaceWithAttributes/attributes}}"] to the
+                [=set/difference=] of
+                |element|["{{SanitizerElementNamespaceWithAttributes/attributes}}"] and
+                |configuration|["{{SanitizerConfig/attributes}}"].
+        1. If |configuration|["{{SanitizerConfig/removeAttributes}}"] [=map/exists=]:
+            1. Set |element|["{{SanitizerElementNamespaceWithAttributes/attributes}}"] to the
+                [=set/intersection=] of
+                |element|["{{SanitizerElementNamespaceWithAttributes/attributes}}"] and
+                |configuration|["{{SanitizerConfig/removeAttributes}}"].
+    1. If |element|["{{SanitizerElementNamespaceWithAttributes/removeAttributes}}"] [=map/exists=]:
+        1. If |configuration|["{{SanitizerConfig/attributes}}"] [=map/exists=]:
+            1. Set |element|["{{SanitizerElementNamespaceWithAttributes/removeAttributes}}"] to the
+                [=set/difference=] of
```

Comment:
All local removeAttributes must be part of the global attributes, so this must an intersection.